internal/controller/object/object.go

```diff
@@ -479,6 +472,103 @@ func getReferenceInfo(ref v1alpha2.Reference) (string, string, string, string) {
 	return apiVersion, kind, namespace, name
 }
 
+func (c *external) checkDeriveFromObject(observed *unstructured.Unstructured) (ready bool) {
+	conditioned := xpv1.ConditionedStatus{}
+	err := fieldpath.Pave(observed.Object).GetValueInto("status", &conditioned)
+	if err != nil {
+		c.logger.Debug("Got error while getting conditions from observed object, setting it as Unavailable", "error", err, "observed", observed)
+		return
+	}
+	if status := conditioned.GetCondition(xpv1.TypeReady).Status; status != v1.ConditionTrue {
+		c.logger.Debug("Observed object is not ready, setting it as Unavailable", "status", status, "observed", observed)
+		return
+	}
+	ready = true
+	return
+}
+
+func (c *external) checkAllConditions(observed *unstructured.Unstructured) (allTrue bool) {
+	conditioned := xpv1.ConditionedStatus{}
+	err := fieldpath.Pave(observed.Object).GetValueInto("status", &conditioned)
+	if err != nil {
+		c.logger.Debug("Got error while getting conditions from observed object, setting it as Unavailable", "error", err, "observed", observed)
+		return
+	}
+	allTrue = len(conditioned.Conditions) > 0
+	for _, condition := range conditioned.Conditions {
+		if condition.Status != v1.ConditionTrue {
+			allTrue = false
+			break
+		}
+	}
+	return
+}
+
+const (
+	msgCelQueryFailedToCompile       = "failed to compile query; %s"
+	msgCelQueryReturnTypeNotBool     = "celQuery does not return a bool type; %s"
+	msgCelQueryFailedToCreateProgram = "failed to create program from the cel query; %s"
+	msgCelQueryFailedToEvalProgram   = "failed to eval the program; %s"
+)
+
+// checkDeriveFromCelQuery will look at the celQuery field and run it as a program, using the observed object as input to
+// evaluate if the object is ready or not
+func (c *external) checkDeriveFromCelQuery(obj *v1alpha2.Object, observed *unstructured.Unstructured) (ready bool, message string) {
+	if obj.Spec.Readiness.CelQuery == "" {
+		c.logger.Debug("cel query is empty")
+		message = "cel query cannot be empty"
+		return ready, message
+	}
```

Comment:
this could be done at the CRD level, to avoid erroring out at runtime 🤔

Reply:
Good idea. Should be a validation. Will add.

Reply:
I've kept this block. I'm worried the cel validation at the CRD level doesn't kick in to all supported versions of kubernetes yet. Anyways always good to guard just in case.